openedx/tutor-contrib-aspects
# User PII dictionary table

Adds `event_sink.user_pii` dictionary table created by joining `external_id` and `user_profile` tables, using the most recent `user_profile` row. This dictionary is cached in memory, refreshing every `ASPECTS_PII_CACHE_LIFETIME` seconds.

Also partitions the `user_profile` (and `external_id`) tables to prepare for user retirement deletions.

Note: A user may have more than one `external_user_id`, and so `event_sink.user_pii` may show more than one row per user.

### Related information

Addresses https://github.com/openedx/tutor-contrib-aspects/issues/449
Depends on https://github.com/openedx/openedx-event-sink-clickhouse/pull/63
Based off https://github.com/openedx/tutor-contrib-aspects/pull/521 which adds the `event_sink.external_id` table.

### Testing instructions

1. Enable Aspects PII data sinks:
   `tutor config save -s ASPECTS_ENABLE_PII=True`
1. Shorten the lifetime of the user_pii dict to 10s to make testing data changes easier:
   `tutor config save -s ASPECTS_PII_CACHE_LIFETIME=10`
1. Apply migrations from this PR:
   `tutor dev do alembic -c "upgrade head"`
1. Login to Superset as a superuser, and navigate to the SQL Lab
1. Check that the `user_pii` table is empty:
   `select * from event_sink.user_pii`
1. Create a new user:
   `tutor dev do createuser any user@example.com`
1. Re-check the user_pii table -- should have an entry for the new user
1. Login to the LMS as the new user and change some of the Account details
1. Re-check the user_pii table -- should show the updated data for the user within 10s.
1. Enrol in a course (this should create an `external_id` row).
1. Re-check the user_pii table -- should show the updated `external_user_id` for the user within 10s.
1. Check migration rollback works:
   `tutor dev do alembic -c "downgrade -3"`

### To do before merge:

- [ ] Merge & tag https://github.com/openedx/openedx-event-sink-clickhouse/pull/63 first
- [ ] Update [openedx-event-sink-clickhouse dependency](https://github.com/openedx/tutor-contrib-aspects/pull/529/files#diff-63394dd6dfdc2f0e274c0aab2b8682130b8a2f3cd85eba391aa52164dde8802aL43)
